torch_np/_ndarray.py

```diff
@@ -340,7 +341,7 @@ def reshape(self, *shape, order="C"):
         # if sh = (1, 2, 3), numpy allows both .reshape(sh) and .reshape(*sh)
         if order != "C":
             raise NotImplementedError
-        tensor = self._tensor.reshape(newshape)
+        tensor = self._tensor.reshape(_helpers.ndarrays_to_tensors(newshape))
```

Comment:
I don't understand this change. `reshape` accepts a tuple of integers as an input.

Reply:
Both torch and numpy accept 0-d arrays in `reshape`

```python
>>> t = torch.arange(25); torch.reshape(t, (torch.as_tensor(5), 5))
tensor([[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24]])

>>> x = np.arange(25); np.reshape(x, (np.asarray(5), 5))
array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])
```

so use of `_helpers.ndarrays_to_tensors` normalise any `tnp.ndarray` inputs to torch tensors before they get passed to `torch.Tensor.reshape()`.

---

... I just noticed that torch is (presumably) just `int()`/`operator.index()`-ing inputs, as this works too

```python
>>> t = torch.arange(25); torch.reshape(t, (tnp.asarray(5), 5))
tensor([[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24]])
```

so this change is only semantically valuable (which might be enough?).

Reply:
So numpy reshape accepts not just tuples and unpacked typles, but also arrays, ugh. 

```
In [72]: a
Out[72]: array_w([1, 2, 3, 4])

In [73]: a._tensor.numpy().reshape(np.array((2, 2)))
Out[73]: 
array([[1, 2],
       [3, 4]])
```

However on main, 0d arrays seem to work already?

```
In [75]: a.reshape(2, tnp.array(2))
Out[75]: 
array_w([[1, 2],
        [3, 4]])
```

Reply:
> However on main, 0d arrays seem to work already?

Probably because torch likes to fallback on `__{int/index}__` of arguments to functions like `reshape()`.

So to recap, I think this change is fine? Just to say due to the rebase, I now do this in `_detail.implementations.reshape()` rather than `ndarray.reshape()`.

Reply:
can you please leave a comment though?

Reply:
> can you please leave a comment though?

Ayup done